File: gherkin/dart/lib/src/pickles/PickleCompiler.dart
import 'package:gherkin/ast.dart';
import 'package:gherkin/extensions.dart';
import 'package:gherkin/language.dart';
import 'package:gherkin/pickles.dart';

class PickleCompiler
{
  final IdGenerator _idGenerator;

  PickleCompiler(this._idGenerator);

  List<Pickle> compile(GherkinDocument gherkinDocument, String uri) {
    var pickles = <Pickle>[];
    var feature = gherkinDocument.feature;
    if (feature.isEmpty) {
      return pickles;
    }

    var language = feature.language;

    compileFeature(pickles, feature, language, uri);
    return pickles;
  }

  void compileFeature(List<Pickle> pickles, Feature feature, String language
      , String uri)
  {
    var tags = feature.tags.toList();
    var featureBackgroundSteps = <Step>[];
    for (var child in feature.children) {
      if (child.background.isNotEmpty) {
        featureBackgroundSteps.addAll(child.background.steps);
      }
      else if (child.rule.isNotEmpty) {
        _compileRule(pickles, child.rule, tags, featureBackgroundSteps, language, uri);
      }
      else {
        var scenario = child.scenario;
        if (scenario.examples.isEmpty) {
          _compileScenario(pickles, scenario, tags, featureBackgroundSteps, language, uri);
        }
        else {
          _compileScenarioOutline(pickles, scenario, tags, featureBackgroundSteps, language, uri);
        }
      }
    }
  }

  void _compileRule(List<Pickle> pickles, Rule rule, List<Tag> tags
      , List<Step> featureBackgroundSteps, String language, String uri)
  {
    final ruleBackgroundSteps = List<Step>.from(featureBackgroundSteps);
    for (var child in rule.children) {
      if (child.background.isNotEmpty) {
        ruleBackgroundSteps.addAll(child.background.steps);
      }
      else {
        var scenario = child.scenario;
        if (scenario.examples.isEmpty) {
          _compileScenario(pickles, scenario, tags, ruleBackgroundSteps, language, uri);
        }
        else {
          _compileScenarioOutline(pickles, scenario, tags, ruleBackgroundSteps, language, uri);
        }
      }
    }
  }

  void _compileScenario(List<Pickle> pickles, Scenario scenario
      , List<Tag> parentTags, List<Step> backgroundSteps, String language
      , String uri)
  {
    final steps = <PickleStep>[];
    if (scenario.steps.isNotEmpty) {
      steps.addAll(_buildPickleSteps(backgroundSteps));
    }

    steps.addAll(_buildPickleSteps(scenario.steps));

    final scenarioTags = <Tag>[];
    scenarioTags.addAll(parentTags);
    scenarioTags.addAll(scenario.tags);

    final pickleTags = _buildPickleTags(scenarioTags);

    var sourceIds = <String>[scenario.id];
    var pickle = Pickle(
        _idGenerator.newId(),
        uri,
        scenario.name,
        language,
        steps,
        pickleTags,
        sourceIds);

    pickles.add(pickle);
  }

  void _compileScenarioOutline(List<Pickle> pickles, Scenario scenario
      , List<Tag> featureTags, List<Step> backgroundSteps, String language
      , String uri) 
  {
    for (final examples in scenario.examples)
    {
      if (examples.tableHeader.isEmpty) {
        continue;
      }
      
      var variableCells = examples.tableHeader.cells;

      for (final valuesRow in examples.tableBody)
      {
        var valueCells = valuesRow.cells;

        var steps = <PickleStep>[];

        if (scenario.steps.isEmpty) {
          final steps = _buildPickleSteps(backgroundSteps);
          steps.addAll(steps);
        }

        var tags = <Tag>[];
        tags.addAll(featureTags);
        tags.addAll(scenario.tags);
        tags.addAll(examples.tags);

        for (var scenarioOutlineStep in scenario.steps) {
          var pickleStep 
            = _buildPickleStep(scenarioOutlineStep, variableCells, valuesRow);

          steps.add(pickleStep);
        }

        var sourceIds = <String>[scenario.id, valuesRow.id];
        final name = _interpolate(scenario.name, variableCells, valueCells);
        final pickleTags = _buildPickleTags(tags);

        var pickle = Pickle(
          _idGenerator.newId(),
          uri,
          name,
          language,
          steps,
          pickleTags,
          sourceIds);
  
        pickles.add(pickle);
      }
    }
  }

  List<PickleStep> _buildPickleSteps(Iterable<Step> steps) {
    final result = <PickleStep>[];
    for (var step in steps) {
      result.add(_buildPickleStep(step));
    }
    return List<PickleStep>.unmodifiable(result);
  }

  PickleStep _buildPickleStep(Step step
      , [Iterable<TableCell> variableCells = const <TableCell>[]
      , TableRow valuesRow = TableRow.empty] )
  {
    var valueCells = valuesRow.isEmpty ? <TableCell>[] : valuesRow.cells;
    var stepText = _interpolate(step.text, variableCells
        , valueCells);

    PickleStepArgument argument = PickleStepArgument.empty;

    if (step.dataTable.isNotEmpty) {
      final dataTable = _buildPickleDataTable(step.dataTable, variableCells
          , valueCells);
      argument = PickleStepArgument(dataTable: dataTable);
    }

    if (step.docString.isNotEmpty) {
      final pickleDocString = _buildPickleDocString(step.docString
          , variableCells, valueCells);
      argument = PickleStepArgument(docString: pickleDocString);
    }

    var pickleStep = PickleStep(_idGenerator.newId(), stepText, argument);

    pickleStep.astNodeIds.add(step.id);

    if(valuesRow.isNotEmpty) {
      pickleStep.astNodeIds.add(valuesRow.id);
    }

    return pickleStep;
  }

  List<PickleTag> _buildPickleTags(List<Tag> tags) {
    var result = <PickleTag>[];
    for (var tag in tags) {
      result.add(buildPickleTag(tag));
    }
    return result;
  }

  PickleTag buildPickleTag(Tag tag) {
    return PickleTag(tag.name, tag.id);
  }

  PickleTable _buildPickleDataTable(DataTable dataTable
      , Iterable<TableCell> variableCells, Iterable<TableCell> valueCells)
  {
    var rows = dataTable.rows;
    var newRows = <PickleTableRow>[];
    for (var row in rows)
    {
      var cells = row.cells;
      var newCells = <PickleTableCell>[];

      for (var cell in cells)
      {
        final value = _interpolate(cell.value, variableCells, valueCells);
        final pickleTableCell = PickleTableCell(value);
        newCells.add( pickleTableCell );
      }

      final pickleTableRow = PickleTableRow(newCells);

      newRows.add(pickleTableRow);
    }

    final pickleTable = PickleTable(newRows);
    return pickleTable;
  }

  PickleDocString _buildPickleDocString(DocString docString
      , Iterable<TableCell> variableCells, Iterable<TableCell> valueCells)
  {
    final content = _interpolate(docString.content, variableCells, valueCells);
    final mediaType = docString.mediaType.isEmpty
        ? Strings.empty
        : _interpolate(docString.mediaType, variableCells, valueCells);

    return PickleDocString(mediaType, content);
  }

  String _interpolate(String name, Iterable<TableCell> variableCells
      , Iterable<TableCell> valueCells) {
    var col = 0;
    for (var variableCell in variableCells) {
      final valueCell = valueCells.elementAt(col++);
      final header = variableCell.value;
      final value = valueCell.value;
      name = name.replaceAll('<' + header + '>', value);
    }
    return name;
  }
}

File: gherkin/java/src/main/java/io/cucumber/gherkin/PickleCompiler.java
package io.cucumber.gherkin;

import io.cucumber.messages.IdGenerator;
import io.cucumber.messages.types.DataTable;
import io.cucumber.messages.types.DocString;
import io.cucumber.messages.types.Examples;
import io.cucumber.messages.types.Feature;
import io.cucumber.messages.types.FeatureChild;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Pickle;
import io.cucumber.messages.types.PickleDocString;
import io.cucumber.messages.types.PickleStep;
import io.cucumber.messages.types.PickleStepArgument;
import io.cucumber.messages.types.PickleTable;
import io.cucumber.messages.types.PickleTableCell;
import io.cucumber.messages.types.PickleTableRow;
import io.cucumber.messages.types.PickleTag;
import io.cucumber.messages.types.Rule;
import io.cucumber.messages.types.RuleChild;
import io.cucumber.messages.types.Scenario;
import io.cucumber.messages.types.Step;
import io.cucumber.messages.types.TableCell;
import io.cucumber.messages.types.TableRow;
import io.cucumber.messages.types.Tag;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static java.util.Collections.unmodifiableList;

class PickleCompiler {

    private final IdGenerator idGenerator;

    public PickleCompiler(IdGenerator idGenerator) {
        this.idGenerator = idGenerator;
    }

    public List<Pickle> compile(GherkinDocument gherkinDocument, String uri) {
        List<Pickle> pickles = new ArrayList<>();
        if (!gherkinDocument.getFeature().isPresent()) {
            return pickles;
        }
        Feature feature = gherkinDocument.getFeature().get();

        String language = feature.getLanguage();

        compileFeature(pickles, feature, language, uri);
        return pickles;
    }

    private void compileFeature(List<Pickle> pickles, Feature feature, String language, String uri) {
        List<Tag> tags = feature.getTags();
        List<Step> featureBackgroundSteps = new ArrayList<>();
        for (FeatureChild featureChild : feature.getChildren()) {
            if (featureChild.getBackground().isPresent()) {
                featureBackgroundSteps.addAll(featureChild.getBackground().get().getSteps());
            } else if (featureChild.getRule().isPresent()) {
                compileRule(pickles, featureChild.getRule().get(), tags, featureBackgroundSteps, language, uri);
            } else if (featureChild.getScenario().isPresent()) {
                Scenario scenario = featureChild.getScenario().get();
                if (scenario.getExamples().isEmpty()) {
                    compileScenario(pickles, scenario, tags, featureBackgroundSteps, language, uri);
                } else {
                    compileScenarioOutline(pickles, scenario, tags, featureBackgroundSteps, language, uri);
                }
            }
        }
    }

    private void compileRule(List<Pickle> pickles, Rule rule, List<Tag> parentTags, List<Step> featureBackgroundSteps, String language, String uri) {
        List<Step> ruleBackgroundSteps = new ArrayList<>(featureBackgroundSteps);

        List<Tag> ruleTags = new ArrayList<>();
        ruleTags.addAll(parentTags);
        ruleTags.addAll(rule.getTags());

        for (RuleChild ruleChild : rule.getChildren()) {
            if (ruleChild.getBackground().isPresent()) {
                ruleBackgroundSteps.addAll(ruleChild.getBackground().get().getSteps());
            } else if (ruleChild.getScenario().isPresent()) {
                Scenario scenario = ruleChild.getScenario().get();
                if (scenario.getExamples().isEmpty()) {
                    compileScenario(pickles, scenario, ruleTags, ruleBackgroundSteps, language, uri);
                } else {
                    compileScenarioOutline(pickles, scenario, ruleTags, ruleBackgroundSteps, language, uri);
                }
            }
        }
    }

    private void compileScenario(List<Pickle> pickles, Scenario scenario, List<Tag> parentTags, List<Step> backgroundSteps, String language, String uri) {
        List<PickleStep> steps = new ArrayList<>();
        if (!scenario.getSteps().isEmpty())
            steps.addAll(pickleSteps(backgroundSteps));

        steps.addAll(pickleSteps(scenario.getSteps()));

        List<Tag> scenarioTags = new ArrayList<>();
        scenarioTags.addAll(parentTags);
        scenarioTags.addAll(scenario.getTags());

        List<String> sourceIds = singletonList(scenario.getId());

        Pickle pickle = new Pickle(
                idGenerator.newId(),
                uri,
                scenario.getName(),
                language,
                steps,
                pickleTags(scenarioTags),
                sourceIds
        );
        pickles.add(pickle);
    }

    private void compileScenarioOutline(List<Pickle> pickles, Scenario scenario, List<Tag> featureTags, List<Step> backgroundSteps, String language, String uri) {
        for (final Examples examples : scenario.getExamples()) {
            if (!examples.getTableHeader().isPresent()) continue;
            List<TableCell> variableCells = examples.getTableHeader().get().getCells();
            for (final TableRow valuesRow : examples.getTableBody()) {
                List<TableCell> valueCells = valuesRow.getCells();

                List<PickleStep> steps = new ArrayList<>();

                if (!scenario.getSteps().isEmpty())
                    steps.addAll(pickleSteps(backgroundSteps));


                List<Tag> tags = new ArrayList<>();
                tags.addAll(featureTags);
                tags.addAll(scenario.getTags());
                tags.addAll(examples.getTags());

                for (Step scenarioOutlineStep : scenario.getSteps()) {
                    PickleStep pickleStep = pickleStep(scenarioOutlineStep, variableCells, valuesRow);

                    steps.add(pickleStep);
                }

                List<String> sourceIds = asList(scenario.getId(), valuesRow.getId());
                Pickle pickle = new Pickle(
                        idGenerator.newId(),
                        uri,
                        interpolate(scenario.getName(), variableCells, valueCells),
                        language,
                        steps,
                        pickleTags(tags),
                        sourceIds
                );

                pickles.add(pickle);
            }
        }
    }

    private PickleTable pickleDataTable(DataTable dataTable, List<TableCell> variableCells, List<TableCell> valueCells) {
        List<TableRow> rows = dataTable.getRows();
        List<PickleTableRow> newRows = new ArrayList<>(rows.size());
        for (TableRow row : rows) {
            List<TableCell> cells = row.getCells();
            List<PickleTableCell> newCells = new ArrayList<>();
            for (TableCell cell : cells) {
                newCells.add(new PickleTableCell(interpolate(cell.getValue(), variableCells, valueCells)));
            }
            newRows.add(new PickleTableRow(newCells));
        }
        return new PickleTable(newRows);
    }

    private PickleDocString pickleDocString(DocString docString, List<TableCell> variableCells, List<TableCell> valueCells) {
        return new PickleDocString(
                docString.getMediaType().isPresent() ? interpolate(docString.getMediaType().get(), variableCells, valueCells) : null,
                interpolate(docString.getContent(), variableCells, valueCells)
        );
    }

    private PickleStep pickleStep(Step step, List<TableCell> variableCells, TableRow valuesRow) {
        List<TableCell> valueCells = valuesRow == null ? Collections.emptyList() : valuesRow.getCells();
        String stepText = interpolate(step.getText(), variableCells, valueCells);

        PickleStepArgument argument = null;
        if (step.getDataTable().isPresent()) {
            argument = new PickleStepArgument(null, pickleDataTable(step.getDataTable().get(), variableCells, valueCells));
        }

        if (step.getDocString().isPresent()) {
            argument = new PickleStepArgument(pickleDocString(step.getDocString().get(), variableCells, valueCells), null);
        }


        List<String> astNodeIds;
        if (valuesRow != null) {
            astNodeIds = Stream.of(singletonList(step.getId()), singletonList(valuesRow.getId()))
                    .flatMap(Collection::stream)
                    .collect(Collectors.toList());

        } else {
            astNodeIds = singletonList(step.getId());
        }

        return new PickleStep(
                argument,
                astNodeIds,
                idGenerator.newId(),
                // TODO: do not pass null - fix this in 1741
                null,
                stepText
        );
    }

    private List<PickleStep> pickleSteps(List<Step> steps) {
        List<PickleStep> result = new ArrayList<>();
        for (Step step : steps) {
            result.add(pickleStep(step));
        }
        return unmodifiableList(result);
    }

    private PickleStep pickleStep(Step step) {
        return pickleStep(step, Collections.emptyList(), null);
    }

    private String interpolate(String name, List<TableCell> variableCells, List<TableCell> valueCells) {
        int col = 0;
        for (TableCell variableCell : variableCells) {
            TableCell valueCell = valueCells.get(col++);
            String header = variableCell.getValue();
            String value = valueCell.getValue();
            name = name.replace("<" + header + ">", value);
        }
        return name;
    }

    private List<PickleTag> pickleTags(List<Tag> tags) {
        List<PickleTag> result = new ArrayList<>();
        for (Tag tag : tags) {
            result.add(pickleTag(tag));
        }
        return result;
    }

    private PickleTag pickleTag(Tag tag) {
        return new PickleTag(tag.getName(), tag.getId());
    }

}
